### DISCUSSION TITLE
Exporting to ANSYS HFSS

### DISCUSSION THREAD

**Discussion by WorkQuoh** [iqm-finland/KQCircuits#102] [General]
Hi, I am having trouble with exporting design to ANSYS with kqcircuits. Using the ansys_export.py to export my design I will get an error which will say: "export_ansys_json() got an unexpected keyword argument 'ansys_tool'". The reason is because this line of code in ansys_export.lym: "json = export_ansys_json(simulation, path, ansys_tool='hfss')  # Alternatively choose ansys_tool='q3d' for Q3D export". However, if you go back to the ansys_export.py you will find this function has no parameter "ansys_tool" in the function. These are the parameters that ansys_export.py has for this function: "def export_ansys_json(simulation: Union[Simulation, CrossSectionSimulation], solution: AnsysSolution, path: Path):". This must have not been updated on the GitHub because there is no way to do export in ANSYS HFSS.

**Comment by qpavsmi** [comment]
Hi! The `ansys_export.lym` macro is a bit outdated. If you open the macro in KLayout's macro editor and modify its contents with following, it should work then:

```
from kqcircuits.klayout_view import KLayoutView
from kqcircuits.simulations.export.ansys.ansys_solution import AnsysHfssSolution, AnsysQ3dSolution
from kqcircuits.simulations.export.simulation_export import copy_content_into_directory
from kqcircuits.simulations.simulation import Simulation
from kqcircuits.simulations.export.ansys.ansys_export import export_ansys_json, export_ansys_bat
from kqcircuits.util.export_helper import create_or_empty_tmp_directory, write_commit_reference_file
from kqcircuits.defaults import TMP_PATH, ANSYS_EXECUTABLE, ANSYS_SCRIPT_PATHS

"""Before using this script, have your project in view as the top cell."""

name = "Dev"  # makes filename Dev.json and Dev.gds
path = TMP_PATH / name # Your path here

top_cell = KLayoutView(current=True).active_cell

simulation = Simulation.from_cell(top_cell, margin=100, name=name)

create_or_empty_tmp_directory(path)
write_commit_reference_file(path)
copy_content_into_directory(ANSYS_SCRIPT_PATHS, path, "scripts")
json = export_ansys_json(simulation, AnsysHfssSolution(), path)  # Alternatively choose AnsysQ3dSolution for Q3D export
exec_file = export_ansys_bat([json], path)
print(f"Simulation exported, execute {exec_file} to run simulation in Ansys")
```

I would recommend to study the method that we support more at this time, which is to define the element in code that you wish to simulate, then running the `kqc sim` utility. You will need [standalone installation](https://iqm-finland.github.io/KQCircuits/developer/standalone.html#basic-installation) to do this.

Here is an example of simulating the [SingleXmons chip](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/kqcircuits/chips/single_xmons.py).

We define a [simulation class](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/kqcircuits/simulations/single_xmons_full_chip_sim.py) that uses this chip and defines ports for it, and removes elements that are not relevant for simulation.

Finally, we have the [simulation script](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/xs1_full_chip_sim.py) where we define Ansys configuration. Not present in this script, but you are also able to sweep different properties of the chip using functions [sweep_simulation](https://github.com/iqm-finland/KQCircuits/blob/600283f4d8d9301eaacbb2d7830256213459d593/klayout_package/python/kqcircuits/simulations/export/simulation_export.py#L104) or [cross_sweep_simulation](https://github.com/iqm-finland/KQCircuits/blob/600283f4d8d9301eaacbb2d7830256213459d593/klayout_package/python/kqcircuits/simulations/export/simulation_export.py#L120). For example the readout resonator lengths (`readout_res_lengths`) could be sweeped for qubits of the chip. Then running command `kqc sim <SIMULATION SCRIPT FILE>` will run simulations for every sweep. This is an advantage over simulating a manually crafted design, which would need to be modified manually for each such sweep.

For single element simulations like [ConcentricTransmon coupling simulation](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/concentrictransmon_couplings_sim.py) you can skip defining a simulation class and just use the function [get_single_element_sim_class](https://github.com/iqm-finland/KQCircuits/blob/600283f4d8d9301eaacbb2d7830256213459d593/klayout_package/python/scripts/simulations/concentrictransmon_couplings_sim.py#L44).

More information on the current methods of running simulations can be found [here](https://iqm-finland.github.io/KQCircuits/user_guide/simulations.html) and feel free to ask me for more info.

**Comment by WorkQuoh** [comment]
Hi thanks for the response! So, I tried that script, and it worked however I ran into other problems as well. For example, I had to change the geometry.py file to allow the import of the geometries since there was a syntax error: 

This is what fixed it for me: 
```
# ["NAME:ChangedProps", ["NAME:Material", "Value:=", f'"{material}"']],  #This causes an error
   ["NAME:ChangedProps", ["NAME:Material", "Value:=", '"{}"'.format(material)]],  #This works
```

This fixes it because fstrings like f"" in python only started in version 3.6 I believe, so you have to use the older .format('') fstring for it not to be a syntax error. I only ran into these issues once I was in the ANSYS HFSS software, so I think there must be some compatibility issues that are causing the problem. Also, I am using the latest version of ANSYS HFSS 2024 version. 

I would be able to import some of my designs and others it would not, and many times port and materials were not assigned either so there seems to be many issues at the moment with getting the simulation started.

Any help will be greatly appreciated. Thanks in Advance.

**Reply by qpavsmi** [reply]
Hi, my apologies. The newer python formatting was introduced by error, and I didn't update this repository with a fix. I have updated the repository now, so fstrings should no longer be a problem in Ansys. If you followed the installation instructions from our documentation, then if you `cd` to the KQCircuits folder and command `git pull`, it should now fix it for you.

When it comes to other issues, please do give more details on those. Again, if it would be possible to implement your designs in code rather than as static design (oas/gds) files, it will make it less difficult down the line to configure the simulation ports correctly.

**Reply by qpavsmi** [reply]
Also sharing the exported simulation files with us will help us figure out the problem faster, if you're comfortable with that of course. You can do so over the email or discord if you wish:
email: pavel.smirnov@meetiqm.com
discord: _pavelsmirnov

**Reply by WorkQuoh** [reply]
I am exporting the Single Xmons as a test. 

I tried the ```git pull``` and this is the new error I get.

![image](https://github.com/user-attachments/assets/1ba53737-bad6-4660-8da7-588a33a3f59f)
![image](https://github.com/user-attachments/assets/ca43b9ac-ddb4-4e02-b06d-77fc78dfee12)

**Reply by qpavsmi** [reply]
Did you run the `kqc sim xs1_full_chip_sim.py` command, or did you place the Single Xmons chip on KLayout editor, then run `ansys_export` macro? If you did the latter, try doing the former, but first set `use_elmer = False` in [xs1_full_chip_sim.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/xs1_full_chip_sim.py)

**Reply by WorkQuoh** [reply]
I did the later like you mentioned. I am now trying the way you suggested, by running xs1_full_chip_sim.py in the Macro development. When I run this code, I get the following error: 'InternalPort' object has no attribute 'etch_width'

**Reply by qpavsmi** [reply]
Can you double check that the code stored locally in your file system matches the code in github? You can do this by running `git status`, it should say `nothing to commit, working tree clean`. Also issuing `git pull` shouldn't hurt.

If you go in your local files to `KQCircuits/klayout_package/python/kqcircuits/simulations/port.py` to the `class InternalPort` there should be `etch_width` as seen [here](https://github.com/iqm-finland/KQCircuits/blob/ddc86d32fecc329e59d8aa57c52cf29e77d7664b/klayout_package/python/kqcircuits/simulations/port.py#L83)

**Reply by WorkQuoh** [reply]
If I do ```git status``` this is what I get.

```
\Documents\KQCircuits>git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   klayout_package/python/kqcircuits/defaults.py
        modified:   klayout_package/python/scripts/macros/export/export_ansys.lym
        modified:   klayout_package/python/scripts/simulations/xs1_full_chip_sim.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        klayout_package/python/kqcircuits/chips/SQUIRLMON
        klayout_package/python/kqcircuits/chips/SQUIRLMON.py
        klayout_package/python/kqcircuits/chips/chip2.py
        klayout_package/python/kqcircuits/elements/chip_frame_mkl.py
        klayout_package/python/kqcircuits/simulations/SQUIRLsim.py
        klayout_package/python/kqcircuits/simulations/test.py
        klayout_package/python/scripts/macros/export/export_ANSYS_2.lym

no changes added to commit (use "git add" and/or "git commit -a")
```

I already did a ```git pull```. If I run a ```git pull``` again this is what it says: 

```
\Documents\KQCircuits>git pull
Already up to date.  
```

So, I think I am good there

When I go here: ```KQCircuits/klayout_package/python/kqcircuits/simulations/port.py``` I do see ```etch_width``` in the ```class InternalPort```

**Reply by qpavsmi** [reply]
If you run `pip list`, for the `kqcircuits` entry does it point to the correct KQCircuits code directory?

Should look something like this:
`kqcircuits                    4.8.4.post9+git.a644a499.dirty C:\Sources\KQCircuits\klayout_package\python`
Where `C:\Sources\KQCircuits` is the path where I have KQCircuits cloned.
If for you it is some unexpected path, cd to KQCircuits code directory and run:
`python -m pip install --no-deps -e klayout_package/python`

**Reply by WorkQuoh** [reply]
If I run ```pip list``` I don't get what you have. I get something like this:

```
annotated-types                       0.7.0
ansys-api-mapdl                       0.5.2
ansys-api-platform-instancemanagement 1.1.0
ansys-mapdl-core                      0.68.4
ansys-mapdl-reader                    0.54.1
ansys-math-core                       0.1.5
ansys-platform-instancemanagement     1.1.2
ansys-pythonnet                       3.1.0rc3
ansys-simai-core                      0.2.3
ansys-tools-path                      0.6.0
appdirs                               1.4.4
attrs                                 24.2.0
certifi                               2024.8.30
cffi                                  1.17.1
charset-normalizer                    3.3.2
click                                 8.1.7
clr-loader                            0.2.6
colorama                              0.4.6
contourpy                             1.3.0
cycler                                0.12.1
defusedxml                            0.7.1
fonttools                             4.53.1
fpdf2                                 2.7.9
geomdl                                5.3.1
grpcio                                1.66.1
idna                                  3.10
importlib_metadata                    8.5.0
jsonschema                            4.23.0
jsonschema-specifications             2023.12.1
kiwisolver                            1.4.7
klayout                               0.29.6
kqcircuits                            4.8.4
matplotlib                            3.9.2
networkx                              3.3
numpy                                 1.26.4
packaging                             24.1
pandas                                2.2.2
pillow                                10.4.0
pip                                   24.2
platformdirs                          4.3.3
plumbum                               1.8.3
pooch                                 1.8.2
protobuf                              4.25.4
psutil                                6.0.0
pyaedt                                0.10.1
pyansys-tools-versioning              0.6.0
pycparser                             2.22
pydantic                              2.8.2
pydantic_core                         2.20.1
pyedb                                 0.27.0
pyiges                                0.3.1
pyparsing                             3.1.4
python-dateutil                       2.9.0.post0
pytz                                  2024.1
pyvista                               0.44.1
pywin32                               306
PyYAML                                6.0.2
referencing                           0.35.1
requests                              2.32.3
requests-toolbelt                     1.0.0
rpds-py                               0.20.0
rpyc                                  6.0.0
Rtree                                 1.3.0
scikit-rf                             1.3.0
scipy                                 1.12.0
scooby                                0.10.0
semver                                3.0.2
setuptools                            75.1.0
six                                   1.16.0
sseclient-py                          1.8.0
tabulate                              0.9.0
toml                                  0.10.2
tomli                                 2.0.1
tqdm                                  4.66.4
typing_extensions                     4.12.2
tzdata                                2024.1
urllib3                               2.2.3
versiontag                            1.2.0
vtk                                   9.3.1
wakepy                                0.9.1
zipp                                  3.20.2
```

**Reply by WorkQuoh** [reply]
I was able to get it to be like this: 
```kqcircuits                            4.8.4.post14+git.ddc86d32.dirty C:\Sources\Documents\KQCircuits\klayout_package\python```
However, I still get the error:

```
KQC SIM EXPORT PATH: C:\Sources\Documents\KQCircuits\tmp\xs1_full_chip_sim_hfss
'InternalPort' object has no attribute 'etch_width'
```

**Reply by qpavsmi** [reply]
Hmm I don't really understand why this would happen if your code is up-to-date in `C:\Sources\Documents\KQCircuits`.

Try opening the python interpreter in command line with command `python`. To exit the python interpreter just type `exit()` and enter.

Now put this code one line at a time into the python interpreter:
```
from kqcircuits.pya_resolver import pya
from kqcircuits.simulations.port import InternalPort
p = InternalPort(1, pya.DPoint(0,0))
p.get_etch_polygon()
```
What you should get at the end is `()`. If you get some error then something is wrong.

Maybe if this is not going anywhere then try re-installing the standalone component of KQCircuits. This will not remove your KQCircuits folder, so your designs will be safe.

1. Make sure you are at `C:\Sources\Documents\KQCircuits`
2. `pip uninstall kqcircuits`
3. `pip install -e "klayout_package/python[tests,simulations]"`
4. Run `kqc sim xs1_full_chip_sim.py`

**Reply by WorkQuoh** [reply]
I opened the command line and tried all the lines of code, and they all worked.

**Reply by qpavsmi** [reply]
So did you reinstall KQCircuits and now running `kqc sim xs1_full_chip_sim.py` works with no issues?

**Reply by WorkQuoh** [reply]
I am trying the part now; however I am getting this issue: 

```
running build_ext
      building 'qutip.cy.inter' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for qutip
Successfully built kqcircuits
Failed to build qutip
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (qutip)
```

when doing:
``` 
pip install -e "klayout_package/python[tests,simulations]"
```

**Reply by qpavsmi** [reply]
OK try following (assuming you have Windows):

```
pip install -r klayout_package/python/requirements/win/requirements.txt klayout_package/python/requirements/win/test-requirements.txt klayout_package/python/requirements/win/sim-requirements.txt
pip install --no-deps -e klayout_package/python
```

**Reply by WorkQuoh** [reply]
I get an error there too:
```
pip install -r klayout_package/python/requirements/win/requirements.txt klayout_package/python/requirements/win/test-requirements.txt klayout_package/python/requirements/win/sim-requirements.txt
ERROR: Invalid requirement: 'klayout_package/python/requirements/win/test-requirements.txt': Expected end or semicolon (after name and no valid version specifier)
    klayout_package/python/requirements/win/test-requirements.txt
                   ^
```

**Reply by qpavsmi** [reply]
Sorry, correct is:

```
pip install -r klayout_package/python/requirements/win/requirements.txt -r klayout_package/python/requirements/win/test-requirements.txt -r klayout_package/python/requirements/win/sim-requirements.txt
```

**Reply by WorkQuoh** [reply]
Trying that method I run into the same issue: 
```
error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for qutip
Failed to build qutip
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (qutip)
```

**Reply by qpavsmi** [reply]
I guess if you don't want to go through the effort of installing Microsoft Visual C++ then you can skip qutip installation. Following hopefully works:

```
pip install klayout numpy scipy<=1.12.0 tqdm versiontag pandas matplotlib scikit-rf
pip install --no-deps -e klayout_package/python
```

**Reply by WorkQuoh** [reply]
I was able to do the installation. However, when I run ```xs1_full_chip_sim.py``` I still get the error ```'InternalPort' object has no attribute 'etch_width'```. But I am getting the error in port.py in line 116.

```
    def get_etch_polygon(self):
        """Returns polygon under which the metal should be etched away"""
        try:
            d = self.signal_location - self.ground_location
            v = (0.5 * self.etch_width / d.length()) * pya.DVector(-d.y, d.x)     <------[ This is where the error is happening ]
            return pya.DPolygon(
                [self.signal_location - v, self.signal_location + v, self.ground_location + v, self.ground_location - v]
            )
        except AttributeError:
            return pya.DPolygon()
```

**Reply by qpavsmi** [reply]
For now maybe just modify this function so it returns `pya.DPolygon()`. If this attribute is not set to InternalPort it should just fallback to returning this empty polygon, I don't understand why it doesn't do that for you

**Reply by qpavsmi** [reply]
Maybe out of curiosity you could tell me what version of python you have?

**Reply by WorkQuoh** [reply]
Is this what you are suggesting for modifying the function: 
```
    def get_etch_polygon(self):
        """Returns polygon under which the metal should be etched away"""
        try:
            #d = self.signal_location - self.ground_location
            #v = (0.5 * self.etch_width / d.length()) * pya.DVector(-d.y, d.x)
            #return pya.DPolygon(
            #    [self.signal_location - v, self.signal_location + v, self.ground_location + v, self.ground_location - v]
            #)
            return pya.DPolygon()

        except AttributeError:
            return pya.DPolygon()
```

**Reply by qpavsmi** [reply]
Yes or more simply:

```
    def get_etch_polygon(self):
        """Returns polygon under which the metal should be etched away"""
        return pya.DPolygon()
        #try:
        #    d = self.signal_location - self.ground_location
        #    v = (0.5 * self.etch_width / d.length()) * pya.DVector(-d.y, d.x)
        #    return pya.DPolygon(
        #        [self.signal_location - v, self.signal_location + v, self.ground_location + v, self.ground_location - v]
        #    )
        #    return pya.DPolygon()

        #except AttributeError:
        #    return pya.DPolygon()
```
If this works and launches the Ansys simulation, you can then refer back to [this comment](https://github.com/iqm-finland/KQCircuits/discussions/102#discussioncomment-10691109) and if you make one of your designs in code, you can make a simulation class similar to `SingleXmonsFullChipSim` where this design gets placed and its ports defined, then finally a script similar to `xs1_full_chip_sim.py` that can be launched to run simulations.

**Reply by WorkQuoh** [reply]
Even when I do that, I still get the same error... of ```'InternalPort' object has no attribute 'etch_width'```

**Reply by qpavsmi** [reply]
Based on your [comment previously](https://github.com/iqm-finland/KQCircuits/discussions/102#discussioncomment-10752117) I can't think of other reason why it fails like this other than you have somewhere a second installation of KQCircuits in your system with outdated code, which does not take `etch_width` into account. You can try again to uninstall kqcircuits with `pip uninstall kqcircuits` then verify with `pip list` that there is no `kqcircuits` entry, then run the `pip install --no-deps -e klayout_package/python` again and then `kqc sim xs1_full_chip_sim.py`. But you have done this already in the past so I doubt it will work.

You can try to export the simulation manually with:
`python klayout_package/python/scripts/simulations/xs1_full_chip_sim.py`, see if that gives same errors. If it doesn't you should get a folder in `KQCircuits/tmp/xs1_full_chip_sim_hfss` where running `simulation.bat` should launch Ansys.

If above gives `etch_width` errors, try running `python` interpreter and input following code lines:
```
from kqcircuits.pya_resolver import pya
from kqcircuits.simulations.single_xmons_full_chip_sim import SingleXmonsFullChipSim
SingleXmonsFullChipSim(pya.Layout(), name="xs1_full_chip_sim", use_ports=True, launchers=True, use_test_resonators=True, n=16, port_size=900)
```
If above also complains about `etch_width`, try following:
```
import kqcircuits
kqcircuits.__file__
```
If this says something other than `C:\Sources\Documents\KQCircuits\klayout_package\python\kqcircuits\__init__.py`, then you have a conflicting kqcircuits installation at that other path.

Sorry this turned into such a contrived bug hunt

**Reply by WorkQuoh** [reply]
I dont know if this will help but this is an error I keep getting when I boot up KQCircuits as well: 

![image](https://github.com/user-attachments/assets/91446cd5-a01f-452e-9f7a-2e54de1bb00b)

**Reply by qpavsmi** [reply]
Try running `python setup_within_klayout.py` and see if that gets rid of the warning. But this I think is not related to the issue you have exporting the simulation

**Reply by WorkQuoh** [reply]
I have been doing this to export the simulation every time through the macros


"You can try to export the simulation manually with:
python klayout_package/python/scripts/simulations/xs1_full_chip_sim.py, see if that gives same errors. If it doesn't you should get a folder in KQCircuits/tmp/xs1_full_chip_sim_hfss where running simulation.bat should launch Ansys."

Which gives me the error every time of the ```etch_width```

I did however try this:
```
from kqcircuits.pya_resolver import pya
from kqcircuits.simulations.single_xmons_full_chip_sim import SingleXmonsFullChipSim
SingleXmonsFullChipSim(pya.Layout(), name="xs1_full_chip_sim", use_ports=True, launchers=True, use_test_resonators=True, n=16, port_size=900)
```

But when I run the code in the macro development it is just running forever, I wonder if it is a run time error now.

**Reply by qpavsmi** [reply]
Try running this in the cmd or powershell utility instead of macro development: type `python` and it will open an interpreter where you can input python code. If it complains about etch_width, try looking up where it reads the kqcircuits module from:

```
from kqcircuits.pya_resolver import pya
from kqcircuits.simulations.single_xmons_full_chip_sim import SingleXmonsFullChipSim
SingleXmonsFullChipSim(pya.Layout(), name="xs1_full_chip_sim", use_ports=True, launchers=True, use_test_resonators=True, n=16, port_size=900)
```

**Reply by WorkQuoh** [reply]
It didn't complain about the ```etch width``` here it just simply ran:
```<kqcircuits.simulations.single_xmons_full_chip_sim.SingleXmonsFullChipSim object at 0x000001397ED266F0>```

**Reply by qpavsmi** [reply]
ok good! So when you type `exit()` so that your cmd terminal exits the python interpreter, can you then using this same cmd terminal run either

`kqc sim xs1_full_chip_sim.py`

or

`python klayout_package/python/scripts/simulations/xs1_full_chip_sim.py`

or will it complain about `etch_width` again?

**Reply by WorkQuoh** [reply]
It runs but it only open klayout and not ANSYS

**Reply by qpavsmi** [reply]
If you close KLayout window it should open ANSYS next if you ran `kqc sim ...`

**Reply by WorkQuoh** [reply]
Awesome, thank you so much it worked!

**Reply by qpavsmi** [reply]
Great, I'm glad it worked out in the end!

I'd recommend cleaning up all of the changes we made in KQC code when debugging the issue. You can use `git status` to see which files were modified and `git diff` to see how they were modified. Probably good to return `port.py` back to what it was, but you can keep `use_elmer = False` in xs1_full_chip_sim.py.
